runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java

```diff
@@ -470,81 +541,514 @@ protected WorkItemCommitRequest persistDirectly(WindmillStateCache.ForKey cache)
+      // Leave pendingDeletes alone. Since we can have multiple values with the same timestamp, we
+      // may still need
+      // overlapping deletes to remove previous entries at this timestamp.
     }
 
     @Override
     public ReadableState<Boolean> isEmpty() {
-      throw new UnsupportedOperationException(
-          String.format("%s is not supported", OrderedListState.class.getSimpleName()));
+      return new ReadableState<Boolean>() {
+        @Override
+        public ReadableState<Boolean> readLater() {
+          WindmillOrderedList.this.readLater();
+          return this;
+        }
+
+        @Override
+        public Boolean read() {
+          return Iterables.isEmpty(WindmillOrderedList.this.read());
+        }
+      };
     }
 
     @Override
     public OrderedListState<T> readLater() {
-      throw new UnsupportedOperationException(
-          String.format("%s is not supported", OrderedListState.class.getSimpleName()));
+      return readRangeLater(null, null);
     }
 
     @Override
-    public OrderedListState<T> readRangeLater(Instant minTimestamp, Instant limitTimestamp) {
-      throw new UnsupportedOperationException(
-          String.format("%s is not supported", OrderedListState.class.getSimpleName()));
+    @SuppressWarnings("FutureReturnValueIgnored")
+    public OrderedListState<T> readRangeLater(
+        @Nullable Instant minTimestamp, @Nullable Instant limitTimestamp) {
+      idTracker.readLater();
+      getFuture(minTimestamp, limitTimestamp);
+      return this;
     }
 
     @Override
     public WorkItemCommitRequest persistDirectly(ForKey cache) throws IOException {
       WorkItemCommitRequest.Builder commitBuilder = WorkItemCommitRequest.newBuilder();
+      TagSortedListUpdateRequest.Builder updatesBuilder =
+          commitBuilder.addSortedListUpdatesBuilder().setStateFamily(stateFamily).setTag(stateKey);
+      try {
+        if (cleared) {
+          // Default range.
+          updatesBuilder.addDeletesBuilder().build();
+          cleared = false;
+        }
+
+        if (!pendingAdds.isEmpty()) {
+          // TODO: Once we start caching data, we should remove this line. We have it here now
+          // because once we persist
+          // added data we forget about it from the cache, so the object is no longer complete.
+          complete = false;
+
+          TagSortedListInsertRequest.Builder insertBuilder = updatesBuilder.addInsertsBuilder();
+          idTracker.add(
+              pendingAdds,
+              (elem, id) -> {
+                try {
+                  ByteString.Output elementStream = ByteString.newOutput();
+                  elemCoder.encode(elem.getValue(), elementStream, Context.OUTER);
+                  insertBuilder.addEntries(
+                      SortedListEntry.newBuilder()
+                          .setValue(elementStream.toByteString())
+                          .setSortKey(elem.getTimestamp().getMillis() * 1000)
```

Comment:
For all of these "millis * 1000" calls, use WindmillTimeUtils.harnessToWindmillTimestamp()

Reply:
done